python-sdk/tests_integration/astro_deploy/master_dag.py

```diff
@@ -94,6 +107,92 @@ def prepare_dag_dependency(task_info, execution_time):
     return _task_list, _dag_run_ids
 
 
+def start_sftp_ftp_services_method():
+    import boto3
+
+    ec2 = boto3.resource("ec2", **AWS_S3_CREDS)
+    instance = ec2.create_instances(
+        ImageId=AMI_ID,
+        MinCount=1,
+        MaxCount=1,
+        InstanceType="t2.micro",
+        SecurityGroupIds=[INBOUND_SECURITY_GROUP_ID],
+    )
+    instance_id = instance[0].instance_id
+    ti = get_current_context()["ti"]
+    ti.xcom_push(key=EC2_INSTANCE_ID_KEY, value=instance_id)
+    while get_instances_status(instance_id) != "running":
+        logging.info("Waiting for Instance to be available in running state. Sleeping for 30 seconds.")
+        time.sleep(30)
+
+
+def get_instances_status(instance_id: str) -> str:
+    """Get the instance status by id"""
+    import boto3
+
+    client = boto3.client("ec2", **AWS_S3_CREDS)
+    response = client.describe_instances(
+        InstanceIds=[instance_id],
+    )
+    instance_details = response["Reservations"][0]["Instances"][0]
+    instance_state: str = instance_details["State"]["Name"]
+    if instance_state == "running":
+        ti = get_current_context()["ti"]
+        ti.xcom_push(key=INSTANCE_PUBLIC_IP, value=instance_details["PublicIpAddress"])
+    return instance_state
+
+
+def create_sftp_ftp_airflow_connection(task_instance: Any) -> None:
+    """
+    Checks if airflow connection exists, if yes then deletes it.
+    Then, create a new sftp_default, ftp_default connection.
+    """
+    sftp_conn = Connection(
+        conn_id="sftp_conn",
+        conn_type="sftp",
+        host=task_instance.xcom_pull(key=INSTANCE_PUBLIC_IP, task_ids=["start_sftp_ftp_services"])[0],
+        login=SFTP_USERNAME,
+        password=SFTP_PASSWORD,
+    )  # create a connection object
+
+    ftp_conn = Connection(
+        conn_id="ftp_conn",
+        conn_type="ftp",
+        host=task_instance.xcom_pull(key=INSTANCE_PUBLIC_IP, task_ids=["start_sftp_ftp_services"])[0],
+        login=FTP_USERNAME,
+        password=FTP_PASSWORD,
+    )  # create a connection object
+
+    session = settings.Session()
+    for conn in [sftp_conn, ftp_conn]:
+        connection = session.query(Connection).filter_by(conn_id=conn.conn_id).one_or_none()
+        if connection is None:
+            logging.info("Connection %s doesn't exist.", str(conn.conn_id))
+        else:
+            session.delete(connection)
+            session.commit()
+            logging.info("Connection %s deleted.", str(conn.conn_id))
+
+        session.add(conn)
+        session.commit()  # it will insert the connection object programmatically.
+        logging.info(f"Connection {conn.conn_id} is created")
```

Comment:
Why are we managing the creation of this connection differently when compared with all the other connections we use?

Reply:
This is done to because we want SFTP/FTP servers while running master dag in astro cloud. Unlike Github actions, we don't have access to services in the local container. 

Also, we are creating ec2 instances on the fly which don't have a static IP, because of which we need to get the host dynamically and create the connections.